src/Tools/dotnet-openapi/src/Commands/BaseCommand.cs

```diff
@@ -0,0 +1,294 @@
+            {
+                var args = new string[] {
+                    "add",
+                    "package",
+                    packageId,
+                    "--version",
+                    version,
+                    "--no-restore"
+                };
+
+                var startInfo = new ProcessStartInfo
+                {
+                    FileName = DotNetMuxer.MuxerPath,
+                    Arguments = string.Join(" ", args),
+                    WorkingDirectory = projectFile.Directory.FullName,
+                    RedirectStandardError = true,
+                    RedirectStandardOutput = true,
+                };
+
+                var process = Process.Start(startInfo);
+
+                if (!process.WaitForExit(20*1000))
+                {
+                    throw new ArgumentException($"Adding package `{packageId}` to `{projectFile.Directory}` took too long.");
+                }
+
+                if (process.ExitCode != 0)
+                {
+                    Error.Write(process.StandardError.ReadToEnd());
+                    Error.Write(process.StandardOutput.ReadToEnd());
+                    Error.Write($"Could not add package `{packageId}` to `{projectFile.Directory}`");
+                    throw new ArgumentException();
+                }
+            }
+        }
+
+        private static IEnumerable<Tuple<string, string>> GetServicePackages(CodeGenerator type)
+        {
+            var name = Enum.GetName(typeof(CodeGenerator), type);
+            var attributes = typeof(Program).Assembly.GetCustomAttributes<AssemblyMetadataAttribute>();
+            var attribute = attributes.Single(a => string.Equals(a.Key, name, StringComparison.OrdinalIgnoreCase));
+
+            var packages = attribute.Value.Split(';', StringSplitOptions.RemoveEmptyEntries);
+            var result = new List<Tuple<string, string>>();
+            foreach (var package in packages)
+            {
+                var tmp = package.Split(':', StringSplitOptions.RemoveEmptyEntries);
+                Debug.Assert(tmp.Length == 2);
+                result.Add(new Tuple<string, string>(tmp[0], tmp[1]));
+            }
+
+            return result;
+        }
+
+        private async Task WriteToFile(Stream content, string destinationPath, bool overwrite)
+        {
+            var destinationExists = File.Exists(destinationPath);
+            if (destinationExists && !overwrite)
+            {
+                await Out.WriteAsync($"Not overwriting existing file '{destinationPath}'.");
+                return;
+            }
+
+            await Out.WriteAsync($"Downloading to '{destinationPath}'.");
+            var reachedCopy = false;
+            try
+            {
+                if (destinationExists)
+                {
+                    // Check hashes before using the downloaded information.
+                    var downloadHash = DownloadFileExtensions.GetHash(content);
```

Comment:
Curious, since all this is doing is comparing hashes, if the local copy was modified after download and is newer, it would be overwritten by the remote copy? Is this the behaviour we want?

Reply:
This makes sense from my perspective: The user added or refreshed an `<OpenApiReference>` to a URL and the chosen location on disk _happened_ to match an existing file. **Definitely** expect the file to be overwritten in the `dotnet openapi refresh` case when the download is different.

Guess we could make downloading over an existing file an error in the `dotnet openapi add` case (but I'm not sure it's necessary). @glennc I defer to you on best behaviour here.

Reply:
No I mean, there's no check of up-to-dateness here. I have the same logic in the gRPC tool but I thought this is a little concerning. I just hope the users are good with their version control and notices and fixes any roll-back that might occur.